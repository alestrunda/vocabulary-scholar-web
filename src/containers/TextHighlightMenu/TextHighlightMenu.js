/* global document */

import React from 'react';
import PropTypes from 'prop-types';
import TextHighlight from '../../components/TextHighlight';
import Dropdown from '../../components/Dropdown';
import { connect } from 'react-redux';
import MenuItem from 'material-ui/MenuItem';
import Snackbar from 'material-ui/Snackbar';
import classNames from 'classnames';
import copy from 'copy-to-clipboard';
import { Link } from 'react-router-dom';
import { appSetSelectedPhrase } from '../../actions/app';
import { queryToWordID } from '../../misc';
import { SNACKBAR_HIDE_DURATION } from '../../constants';

export class TextHighlightMenu extends React.Component {
  constructor(props) {
    super(props);
    this.rootRef = React.createRef();
  }

  state = {
    isMenuOpened: false,
    snackBarCopyOpened: false,
    snackBarCopyText: '',
  };

  componentDidMount() {
    document.addEventListener('click', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClickOutside);
  }

  handleClickOutside = e => {
    //close the menu when clicked outside
    if (this.rootRef.current.contains(e.target)) return;
    this.setState({
      isMenuOpened: false,
    });
  };

  onHighlight = (start, end) => {
    if (start === end) return;
    this.setState({
      isMenuOpened: true,
      start,
      end,
    });
  };

  handleSnackBarCopyClose = () => {
    this.setState({
      snackBarCopyOpened: false,
    });
  };

  handleCopyClick = () => {
    const wasCopied = copy(this.getSelectedString());
    this.setState({
      snackBarCopyOpened: true,
      snackBarCopyText: wasCopied ? 'Text copied to clipboard' : 'Copy failed',
    });
  };

  getSelectedString() {
    return this.props.children.substring(this.state.start, this.state.end);
  }

  handleAddPhraseClick = () => {
    this.props.onPhraseSelected(this.getSelectedString());
  };

  render() {
    const { isPhraseDisabled, children } = this.props;
    return (
      <React.Fragment>
        <div
          className={classNames('el-relative', this.props.className)}
          ref={this.rootRef}
          style={this.props.style}
        >
          <TextHighlight onHighlight={this.onHighlight}>
            {children}
          </TextHighlight>
          {this.state.isMenuOpened && (
            <Dropdown>
              <Link
                to={`/word/${queryToWordID(
                  children.substring(this.state.start, this.state.end)
                )}`}
              >
                <MenuItem
                  style={{ minHeight: 40, lineHeight: '38px' }}
                  primaryText="Search"
                />
              </Link>
              {!isPhraseDisabled && (
                <MenuItem
                  style={{ minHeight: 40, lineHeight: '38px' }}
                  onClick={this.handleAddPhraseClick}
                  primaryText="Add as phrase"
                />
              )}
              <MenuItem
                style={{ minHeight: 40, lineHeight: '38px' }}
                onClick={this.handleCopyClick}
                primaryText="Copy"
              />
            </Dropdown>
          )}
        </div>
        <Snackbar
          open={this.state.snackBarCopyOpened}
          message={this.state.snackBarCopyText}
          autoHideDuration={SNACKBAR_HIDE_DURATION}
          onRequestClose={this.handleSnackBarCopyClose}
        />
      </React.Fragment>
    );
  }
}

TextHighlightMenu.propTypes = {
  className: PropTypes.string,
  children: PropTypes.string,
  isPhraseDisabled: PropTypes.bool,
  onPhraseSelected: PropTypes.func.isRequired,
  style: PropTypes.object,
};

const mapDispatchToProps = dispatch => ({
  onPhraseSelected: phrase => dispatch(appSetSelectedPhrase(phrase)),
});

export default connect(
  null,
  mapDispatchToProps
)(TextHighlightMenu);
