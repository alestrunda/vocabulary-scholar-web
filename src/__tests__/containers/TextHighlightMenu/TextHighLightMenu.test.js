import React from 'react';
import { shallow } from 'enzyme';

import { TextHighlightMenu } from '../../../containers/TextHighlightMenu/TextHighlightMenu';

it('renders correctly', () => {
  const onPhraseSelected = jest.fn();
  const root = shallow(
    <TextHighlightMenu onPhraseSelected={onPhraseSelected} />
  );
  expect(root).toMatchSnapshot();
});
