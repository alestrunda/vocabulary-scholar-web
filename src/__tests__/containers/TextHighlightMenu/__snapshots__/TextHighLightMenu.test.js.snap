// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextHighlightMenu
    onPhraseSelected={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="el-relative"
        >
          <TextHightlight
            classNameHighlighted="text-highlight"
            onHighlight={[Function]}
          >
            
          </TextHightlight>
        </div>,
        <Snackbar
          autoHideDuration={4000}
          message=""
          onRequestClose={[Function]}
          open={false}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextHightlight
              classNameHighlighted="text-highlight"
              onHighlight={[Function]}
            >
              
            </TextHightlight>,
            false,
          ],
          "className": "el-relative",
          "style": undefined,
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "",
              "classNameHighlighted": "text-highlight",
              "onHighlight": [Function],
            },
            "ref": null,
            "rendered": "",
            "type": [Function],
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "autoHideDuration": 4000,
          "message": "",
          "onRequestClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="el-relative"
          >
            <TextHightlight
              classNameHighlighted="text-highlight"
              onHighlight={[Function]}
            >
              
            </TextHightlight>
          </div>,
          <Snackbar
            autoHideDuration={4000}
            message=""
            onRequestClose={[Function]}
            open={false}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextHightlight
                classNameHighlighted="text-highlight"
                onHighlight={[Function]}
              >
                
              </TextHightlight>,
              false,
            ],
            "className": "el-relative",
            "style": undefined,
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "",
                "classNameHighlighted": "text-highlight",
                "onHighlight": [Function],
              },
              "ref": null,
              "rendered": "",
              "type": [Function],
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoHideDuration": 4000,
            "message": "",
            "onRequestClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
